scheds/rust/scx_spark/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1961 @@
+
+/*
+ * Maximum possible CPU number.
+ */
+static u64 nr_cpu_ids = 1;
+
+/*
+ * Runtime throttling.
+ *
+ * Throttle the CPUs by injecting @throttle_ns idle time every @slice_max.
+ */
+const volatile u64 throttle_ns;
+static volatile bool cpus_throttled;
+
+static inline bool is_throttled(void)
+{
+	return READ_ONCE(cpus_throttled);
+}
+
+static inline void set_throttled(bool state)
+{
+	WRITE_ONCE(cpus_throttled, state);
+}
+
+/*
+ * Exit information.
+ */
+UEI_DEFINE(uei);
+
+/*
+ * GPU task detection maps.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, u32);
+	__uint(max_entries, MAX_GPU_TASK_PIDS);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} gpu_tgid SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, u32);
+	__uint(max_entries, MAX_GPU_TASK_PIDS);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} gpu_tid SEC(".maps");
+
+/*
+ * Workload type detection maps.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, struct workload_info);
+	__uint(max_entries, MAX_WORKLOAD_PIDS);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} workload_tgid SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, struct workload_info);
+	__uint(max_entries, MAX_WORKLOAD_PIDS);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} workload_tid SEC(".maps");
+
+/*
+ * Mask of CPUs that the scheduler can use until the system becomes saturated,
+ * at which point tasks may overflow to other available CPUs.
+ */
+private(BPFLAND) struct bpf_cpumask __kptr *primary_cpumask;
+
+/*
+ * Mask of Big (performance) CPUs.
+ */
+private(BPFLAND) struct bpf_cpumask __kptr *big_cpumask;
+
+/*
+ * Mask of Little (energy-efficient) CPUs.
+ */
+private(BPFLAND) struct bpf_cpumask __kptr *little_cpumask;
+
+/*
+ * Mask of Turbo (performance) CPUs.
+ */
+private(BPFLAND) struct bpf_cpumask __kptr *turbo_cpumask;
+
+/*
+ * DSQ dispatch mode.
+ */
+const volatile u32 dsq_mode = 0;
+
+/*
+ * Shared DSQ ID (used when dsq_mode == DSQ_MODE_SHARED).
+ */
+#define SHARED_DSQ_ID 0
+
+/*
+ * Current global vruntime.
+ */
+static u64 vtime_now;
+
+/*
+ * Timer used to inject idle cycles when CPU throttling is enabled.
+ */
+struct throttle_timer {
+	struct bpf_timer timer;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, u32);
+	__type(value, struct throttle_timer);
+} throttle_timer SEC(".maps");
+
+/*
+ * Per-CPU context.
+ */
+struct cpu_ctx {
+	u64 tot_runtime;
+	u64 prev_runtime;
+	u64 last_running;
+	u64 perf_lvl;
+	struct bpf_cpumask __kptr *l3_cpumask;
+	bool has_active_kthread;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, struct cpu_ctx);
+	__uint(max_entries, 1);
+} cpu_ctx_stor SEC(".maps");
+
+
+/*
+ * Detect workload type based on process name and behavior patterns.
+ */
+static void detect_workload_type(const struct task_struct *p, u32 pid, u32 tid) {
```

Comment:
you may eventually want this function to be a bpf global function, see #477 for reference